SnowplowTracker/SnowplowTracker/Emitter.cs

```diff
@@ -0,0 +1,490 @@
+		/// </summary>
+		private void Emit(object state) {
+			sending = true;
+			List<EventRow> events = eventStore.GetDescEventRange (sendLimit);
+
+			// If some events were found start sending!
+			if (events.Count > 0) {
+				Log.Debug ("Emitter: Event count: " + events.Count);
+				List<RequestResult> results = SendRequests (events);
+				events = null;
+
+				int success = 0;
+				int failure = 0;
+
+				foreach (RequestResult result in results) {
+					if (result.success) {
+						foreach (int id in result.rowIds) {
+							eventStore.DeleteEvent (id);
+						}
+						success += result.rowIds.Count;
+					} else {
+						failure += result.rowIds.Count;
+					}
+				}
+
+				Log.Debug (" + Successful: " + success);
+				Log.Debug (" + Failure: " + failure);
+
+				if (failure > 0 && success == 0) {
+					// All events failed; sleep for 10 seconds and then retry.
+					// - Sleep prevents thrashing of database in situations where nothing can send.
+					Log.Error ("Emitter: All events failed to send; pausing emitter for ten seconds.");
+					sending = false;
+					if (emitterTimer != null) {
+						emitterTimer.Change (FAIL_INTERVAL, Timeout.Infinite);
+					}
+				} else {
+					if (emitterTimer != null) {
+						Log.Debug ("Emitter: All events sent successfully; checking for more...");
+						Emit (null);
+					}
+				}
+			} else {
+				Log.Debug ("Emitter: No more events to send; waiting for more.");
+				sending = false;
+				lock (emitLock) {
+					Monitor.Wait(emitLock);
+				}
+				if (emitterTimer != null) {
+					Emit (null);
+				}
+			}
+		}
+
+		// --- Event Senders
+
+		/// <summary>
+		/// From a range of event rows will construct a list of RequestResult objects
+		/// </summary>
+		/// <returns>The results of sending all requests</returns>
+		/// <param name="eventRows">Event rows from the database</param>
+		private List<RequestResult> SendRequests(List<EventRow> eventRows) {
+			if (httpMethod == HttpMethod.GET) {
+				return HttpGet (eventRows);
+			} else {
+				return HttpPost (eventRows);
+			}
+		}
+
+		/// <summary>
+		/// Sends all events as GET requests on background threads
+		/// </summary>
+		/// <returns>The get.</returns>
+		/// <param name="eventRows">Event rows.</param>
+		private List<RequestResult> HttpGet(List<EventRow> eventRows) {
+			List<ReadyRequest> requests = new List<ReadyRequest>();
+			List<RequestResult> results = new List<RequestResult>();
+
+			// Send each row as an individual GET Request
+			foreach (EventRow eRow in eventRows) {
+				TrackerPayload payload = eRow.GetPayload();
+				long byteSize = payload.GetByteSize() + POST_STM_BYTES;
+				bool oversize = byteSize > byteLimitGet;
+				Log.Verbose("Emitter: Sending GET with byte-size: " + byteSize);
+				requests.Add(getGETRequest(payload.GetDictionary(), eRow.GetRowId(), oversize));
+			}
+
+			// Wait for the results of each request
+			foreach (ReadyRequest request in requests) {
+				request.WaitForResult();
```

Comment:
At the moment this method calls `Thread.sleep(1)` in a loop to wait for the results of the requests. The UnityHTTP class has a Send method with a callback parameter - is it possible to use the callback to explicitly communicate between threads and avoid the sleep loop? I'm thinking this would involve the C# equivalent of a [Java CountDownLatch](http://docs.oracle.com/javase/7/docs/api/java/util/concurrent/CountDownLatch.html). The latch's count would initialize with value equal to the number of requests to wait for, each callback would decrement the latch's count, and the HttpGet thread would wait for the count to be 0.

But perhaps the request library's callbacks aren't reliable enough for this to work (e.g. they don't fire when the request fails)?